internal/executor/instance/prebuilt.go

```diff
@@ -0,0 +1,177 @@
+package instance
+
+import (
+	"archive/tar"
+	"bufio"
+	"context"
+	"encoding/json"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/sirupsen/logrus"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type PrebuiltInstance struct {
+	Image      string
+	Dockerfile string
+	Arguments  map[string]string
+}
+
+func CreateTempArchive(dir string) (string, error) {
+	tmpFile, err := ioutil.TempFile("", "cirrus-prebuilt-archive-")
+	if err != nil {
+		return "", err
+	}
+
+	archive := tar.NewWriter(tmpFile)
+
+	if err := filepath.Walk(dir, func(path string, fileInfo os.FileInfo, err error) error {
+		// Handle possible error that occurred when reading this directory entry information
+		if err != nil {
+			return err
+		}
+
+		// We clearly don't want any directories here (because tar)
+		// and probably not interested in special files for now
+		if !fileInfo.Mode().IsRegular() {
+			return nil
+		}
+
+		header, err := tar.FileInfoHeader(fileInfo, fileInfo.Name())
+		if err != nil {
+			return err
+		}
+
+		// Since os.FileInfo doesn't contain the full path to a file
+		// we need to manually update the Name field in the header
+		relPath, err := filepath.Rel(dir, path)
+		if err != nil {
+			return err
+		}
+		header.Name = relPath
+
+		// Write file header
+		if err := archive.WriteHeader(header); err != nil {
+			return err
+		}
+
+		// Write file contents
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		if _, err := io.Copy(archive, file); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return "", err
+	}
+
+	if err := archive.Close(); err != nil {
+		return "", err
+	}
+
+	if err := tmpFile.Close(); err != nil {
+		return "", err
+	}
+
+	return tmpFile.Name(), nil
+}
+
+func (prebuilt *PrebuiltInstance) Run(ctx context.Context, config *RunConfig) error {
+	logger := config.Logger
+	if logger == nil {
+		logger = logrus.New()
+		logger.Out = ioutil.Discard
+	}
+
+	// Create a Docker client
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+
+	// Create an archive with the build context
+	archivePath, err := CreateTempArchive(config.ProjectDir)
+	if err != nil {
+		return err
+	}
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return err
+	}
+	if err := os.Remove(archivePath); err != nil {
```

Comment:
Why is it removed right away?

Reply:
Because you can remove the file right away on Unix-like OS'es and keep working with it. Why would we wait, if it wouldn't be needed in the future anyway? But I'm not sure about Windows yet.

Reply:
But wouldn't Docker Client work with it and send it as context? This Unix behaviour is non intuitive that you can open a file, remove it and still work with it. What will the `err` be? How it can be sure file will be removed. Pretty confusing IMO 😅 Maybe just switch to `defer os.Remove(archivePath)`?

Reply:
See https://github.com/cirruslabs/cirrus-cli/pull/33/commits/056dbde64713a3e6a033faf06693f4a749608d66.